### PR TITLE
[ENH] Enable temporal importance curves for DrCIFClassifier by supporting BaseDecisionTree

### LINKED ISSUE #3140: [DOC] Temporal imprortance curves for DrCIF
### Describe the issue linked to the documentation

the classifier DrCIF has a function temporal_importance_curves

https://www.aeon-toolkit.org/en/latest/api_reference/auto_generated/aeon.classification.interval_based.DrCIFClassifier.html#aeon.classification.interval_based.DrCIFClassifier.temporal_importance_curves

it would be nice to add an example of using it to the notebook for interval based classifiers

https://www.aeon-toolkit.org/en/latest/examples/classification/interval_based.html

### Suggest a potential alternative/fix

_No response_

### PR DESCRIPTION
**Pull request by Nithurshen**
#### Reference Issues/PRs

Fixes #3140 

#### What does this implement/fix? Explain your changes.

This Pull Request fixes a `ValueError` in `DrCIFClassifier.temporal_importance_curves()` that occurred because DrCIF defaults to using `sklearn.tree.DecisionTreeClassifier`, which is a subclass of `BaseDecisionTree`, but the base function only explicitly supported `aeon.classification.sklearn.ContinuousIntervalTree`.

The fix involves modifying `BaseIntervalForest.temporal_importance_curves` to:

1.  **Update the base estimator check** to explicitly allow `sklearn.tree.BaseDecisionTree`.
2.  **Implement the Information Gain (Gains) calculation** for `BaseDecisionTree` by inspecting its internal structure (`tree_.tree_`). The gain is calculated as the normalized reduction in impurity for each split node.
3.  Ensure the final function output is explicitly converted to **lists** for API consistency, resolving a subsequent test failure related to `zip()` returning tuples.

This restores the desirable feature of generating temporal importance curves for `DrCIFClassifier` when using its default settings.

#### Does your contribution introduce a new dependency? If yes, which one?

No

#### Any other comments?

A new test case, `test_drcif_temporal_importance_curves`, was added to confirm the fix and ensure future compatibility.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py [hunk 1]
```diff
@@ -259,3 +261,44 @@ def test_interval_features():
 
     assert est._interval_function == [True]
     assert est._interval_transformer == [True]
+
+
+def test_drcif_temporal_importance_curves():
+    """Testing DrCIFClassifierto execute temporal_importance_curves."""
+    X, y = load_classification("GunPoint", split="TRAIN")
+    X_small, y_small = X[:20], y[:20]
+    n_timepoints = X_small.shape[-1]
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py:274]: Can we test both base classifier types? Also it would be good to have some expected results (it can just be one or two curves) just in case they change the internals of the BaseDecisionTree and we need to know.

File: aeon/base/_estimators/interval_based/base_interval_forest.py [hunk 2]
```diff
@@ -1157,18 +1157,45 @@ def temporal_importance_curves(
...
+                    )
+                    * impurity_left
+                    - (
+                        tree.n_node_samples[tree.children_right[internal_nodes]]
+                        / n_samples_node
+                    )
+                    * impurity_right
+                )
+            else:
+                continue
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/base_interval_forest.py]: do we need to continue here?

File: aeon/base/_estimators/interval_based/base_interval_forest.py [hunk 3]
```diff
@@ -1157,18 +1157,45 @@ def temporal_importance_curves(
...
+                    - (
+                        tree.n_node_samples[tree.children_left[internal_nodes]]
+                        / n_samples_node
+                    )
+                    * impurity_left
+                    - (
+                        tree.n_node_samples[tree.children_right[internal_nodes]]
+                        / n_samples_node
+                    )
+                    * impurity_right
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/base_interval_forest.py:1194]: Could you run me through this a bit, I am not familiar with the internals of the BaseDecisionTree. Does this gather the information gain for each node? I do not see any looping or recursion

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
Looks mostly good, but could you please try plotting the output and compare to the original.

File: aeon/base/_estimators/interval_based/base_interval_forest.py (hunk 3 above)
Comment (Nithurshen) [reply aeon/base/_estimators/interval_based/base_interval_forest.py:1194]: `BaseDecisionTree` stores the tree structure in a low-level Tree object. Instead of recursing through nodes, we use `tree.feature >= 0` to create a mask of all internal nodes at once. The arrays `impurity`, `children_left`, and `n_node_samples` are indexed by node ID.

By using NumPy indexing (e.g., `tree.impurity[tree.children_left[internal_nodes]]`), we calculate the impurity decrease for every split in the tree simultaneously across the entire array. This is mathematically equivalent to iterating through each node but significantly faster.

File: aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py (hunk 1 above)
Comment (Nithurshen) [reply aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py:274]: Done

File: aeon/base/_estimators/interval_based/base_interval_forest.py (hunk 2 above)
Comment (Nithurshen) [reply aeon/base/_estimators/interval_based/base_interval_forest.py]: I have removed it.

**Comment by Nithurshen** [comment]
> Looks mostly good, but could you please try plotting the output and compare to the original.

image

File: aeon/base/_estimators/interval_based/base_interval_forest.py [hunk 4]
```diff
@@ -1157,18 +1156,44 @@ def temporal_importance_curves(
...
+                internal_nodes = np.where(tree.feature >= 0)[0]
+                splits = tree.feature[internal_nodes]
+                impurity = tree.impurity[internal_nodes]
+                impurity_left = tree.impurity[tree.children_left[internal_nodes]]
+                impurity_right = tree.impurity[tree.children_right[internal_nodes]]
+                n_samples_node = tree.n_node_samples[internal_nodes]
+                n_samples_total = self.n_cases_
+                gains = (n_samples_node / n_samples_total) * (
+                    impurity
+                    - (
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/base_interval_forest.py:1185]: what is the purpose of the `(n_samples_node / n_samples_total) *` here? I assume this downweights deeper nodes?

File: aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py [hunk 5]
```diff
@@ -259,3 +261,35 @@ def test_interval_features():
...
+        n_estimators=5,
+        random_state=42,
+    )
+    dr.fit(X_small, y_small)
+
+    names, curves = dr.temporal_importance_curves()
+
+    assert isinstance(names, list)
+    assert isinstance(curves, list)
+    assert len(names) == len(curves) > 0
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py:287]: \> 0?

File: aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py [hunk 6]
```diff
@@ -259,3 +261,35 @@ def test_interval_features():
...
+
+    curve = curves[0]
+    assert isinstance(curve, np.ndarray)
+    assert curve.ndim == 1
+    assert len(curve) == n_timepoints
+    assert np.all(curve >= 0)
+
+    assert np.any(curve > 0), "Importance curve should contain non-zero values"
```
Comment (MatthewMiddlehurst) [inline aeon/base/_estimators/interval_based/tests/test_base_interval_forest.py:295]: if the feature is never used it would make an all 0 curve right?

**Comment by MatthewMiddlehurst** [comment]
Generally the new plot looks good though, thanks.
